### PR TITLE
Feature/151785913/reset password

### PR DESCRIPTION
**Pull request by nebanat**
#### What does this PR do?
Users can now reset their password, in the case if they forget their password
#### Description of Task to be completed?
- When a user clicks on the forgot password link
- User sees a page where he/she can enter their email they use in registering for the CFH game
- A password link is generated with reset token that expires in 1 hour  and sent to their emails
-  When user clicks on the reset they are redirected to where they will reset their password
- PR checks if token is valid and has not expired
#### How should this be manually tested?
When user clicks on the forgot password link and tried to reset their password
#### Any background context you want to provide?
Used Nodemailer to send reset link and random string to generate token
#### What are the relevant pivotal tracker stories?
- **Story Id:** 151785913
- **Story type:** Feature
- **Story:** Users should be able to reset their password 
#### Screenshots (if appropriate)
None
#### Questions:
None

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
   });
 };
 
+exports.password = (req, res) => {
+  const { email, resetLink, resetMessage } = req.body;
+  // checks if email field is not empty
+  if (!email || email.trim() === '') {
+    return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: change this to a 400 status

File: app/controllers/users.js [hunk 2]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
...
+      message: 'Please enter a valid Email address '
+    });
+  }
+
+  // check if resetLink or resetMessage is not empty
+  if (!resetLink || !resetMessage) {
+    return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: change this to a 500 status

File: app/controllers/users.js [hunk 3]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
...
+        });
+      });
+    });
+};
+
+exports.resetPassword = (req, res) => {
+  const { resetToken, password } = req.body;
+  // checks if there is a token
+  if (!password) {
+    return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: 400

File: app/controllers/users.js [hunk 4]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
...
+      message: 'Please enter password'
+    });
+  }
+  // checks if there is a token
+  if (!resetToken) {
+    return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: 400

File: app/controllers/users.js [hunk 5]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
...
+    .exec((err, user) => {
+      if (!user) {
+        return res.status(404).send({
+          message: 'No User associated with this Token'
+        });
+      }
+
+      // checks if the token has expired
+      if (Date.now() > user.resetPassExpiry) {
+        return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: 401

File: app/controllers/users.js [hunk 6]
```diff
@@ -75,6 +77,130 @@ exports.sendInviteEmail = (req, res) => {
...
+        });
+      }
+      user.hashed_password = user.encryptPassword(password);
+      // clear token and expiry time
+      user.resetToken = '';
+      user.resetPassExpiry = '';
+
+      user.save((err) => {
+        if (err) {
+          return res.status(402).send({
```
Comment (dinobi) [inline app/controllers/users.js]: 500

**Review by dinobi — CHANGES_REQUESTED** [review]

File: app/controllers/users.js (hunk 6 above)
Comment (nebanat) [reply app/controllers/users.js]: alright I will change it

File: app/controllers/users.js (hunk 5 above)
Comment (nebanat) [reply app/controllers/users.js]: ok

File: app/controllers/users.js (hunk 4 above)
Comment (nebanat) [reply app/controllers/users.js]: ok

File: app/controllers/users.js (hunk 3 above)
Comment (nebanat) [reply app/controllers/users.js]: ok

File: app/controllers/users.js (hunk 2 above)
Comment (nebanat) [reply app/controllers/users.js]: ok

**Review by enodi — APPROVED** [review]

**Review by ibonly — APPROVED** [review]

**Review by dinobi — APPROVED** [review]

**Pull request merged**